### PR TITLE
Intake subsystem

### PR DESCRIPTION
**Pull request by icai0**
for tracking or something idk ask luke

### CODE REVIEW

File: physics.py [hunk 1]
```diff
@@ -141,6 +141,32 @@ def update(self, limelightPose: Pose2d, targetPose: Pose2d) -> None:
...
+        if (
+            constants.kIntakeCameraMinHorizontalFOV.radians()
+            < ballAngle.radians()
+            < constants.kIntakeCameraMaxHorizontalFOV.radians()
+        ):
+            ballValid = True
+            ballDistance = intakeCameraPose.translation().distance(
+                ballPose.translation()
+            )
+            self.photonvisionNetworkTable.putNumber(constants.kPhotonvisionTargetSimDistanceKey, ballDistance)
```
Comment (lmaxwell24) [inline physics.py:164]: photonvision doesnt export distance I thought?
you need to export the vertical offset and then within the vision subsystem calculate the actual distance

File: subsystems/intakesubsystem.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        self.rightSolenoid = Solenoid(PneumaticsModuleType.REVPH, 1)
+        self.rightSolenoid.set(False)
+
+        if RobotBase.isReal():
+            pass
+        else:
+            self.intakeMotor = PWMVictorSPX(constants.kSimIntakeMotorPort)
+
+    def toggleIntake(self) -> None:
+        self.intakeDeployed = not self.intakeDeployed
```
Comment (lmaxwell24) [inline subsystems/intakesubsystem.py:27]: these should actually move the solenoids, I recommend a general "deployed" getter function that uses if both of them are extended

File: subsystems/intakesubsystem.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+from wpilib import PWMVictorSPX, PneumaticHub, PneumaticsModuleType, RobotBase, Solenoid
+import constants
+
+
+class IntakeSubsystem(SubsystemBase):
+    def __init__(self) -> None:
+        SubsystemBase.__init__(self)
+        self.setName(__class__.__name__)
+        self.intakeDeployed = False  # default to intake retracted and not reverseed
+        self.intakeReversed = False
```
Comment (lmaxwell24) [inline subsystems/intakesubsystem.py:11]: seems unused -  the subsystem should handle all subsystem variables and other commands just call to those functions

File: subsystems/visionsubsystem.py [hunk 4]
```diff
@@ -117,6 +119,106 @@ def reset(self) -> None:
...
+    def update(self) -> None:
+        NetworkTables.initialize()
+        self.photonvisionNetworkTable = NetworkTables.getTable(
+            constants.kPhotonvisionNetworkTableName
+        )
+
+        if self.photonvisionNetworkTable.getBoolean(
+            constants.kPhotonvisionTargetValidKey, False
+        ):
+            if RobotBase.isReal():
```
Comment (lmaxwell24) [inline subsystems/visionsubsystem.py:142]: your general implementation of the system should NOT rely on if the robot is real - simulation should handle making the correct values

File: subsystems/visionsubsystem.py [hunk 5]
```diff
@@ -117,6 +119,106 @@ def reset(self) -> None:
...
+                ballAngleYToCamera = Rotation2d.fromDegrees(
+                    self.photonvisionNetworkTable.getValue(
+                        constants.kPhotonvisionTargetVerticalAngleKey, 0
+                    )
+                )
+
+                distanceToCamera = (
+                    Rotation2d.tan(
+                        ballAngleYToCamera.radians()
+                        + constants.kIntakeCameraMaxVerticalFOV.radians() / 2
```
Comment (lmaxwell24) [inline subsystems/visionsubsystem.py:152]: is this assuming a set rotation on the camera to begin with? if so it should be a constant

File: subsystems/visionsubsystem.py [hunk 6]
```diff
@@ -117,6 +119,106 @@ def reset(self) -> None:
...
+                    )
+                    * constants.kIntakeCameraHeightInMeters
+                )
+            else:
+                distanceToCamera = self.photonvisionNetworkTable.getValue(
+                    constants.kPhotonvisionTargetSimDistanceKey, float("inf")
+                )
+
+            if not constants.kIsIntakeCameraCentered:
```
Comment (lmaxwell24) [inline subsystems/visionsubsystem.py:161]: given current design I think it should be safe to say that the camera WILL be centered horizontally (allowing the largest possible space on each side to see)
in the event that it isnt the offset should make is such that the differences zero out and result in no need for this if statement

File: commands/reverseintake.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        self.setName(__class__.__name__)
+        self.intake = intake
+        self.addRequirements(intake)
+
+    def initialize(self) -> None:
+        print(f"Command: {self.getName()}")
+        self.intake.reverseIntake()
+
+        if self.intake.isIntakeDeployed():
+            self.intake.runIntake(self.intake.intakeReversed)
```
Comment (lmaxwell24) [inline commands/reverseintake.py:18]: I noted this below - it is generally some iffy practice to take values directly from classes, instead these should be handled within the subsystem itself and you should be calling onto those values

File: constants.py [hunk 8]
```diff
@@ -193,6 +193,13 @@
 kLimelightTargetValidKey = "tv"
 kLimelightTargetHorizontalAngleKey = "tx"
 
+# Photonvision
+kPhotonvisionNetworkTableName = "photonvision"
+kPhotonvisionTargetValidKey = "hasTarget"
+kPhotonvisionTargetHorizontalAngleKey = "targetYaw"
+kPhotonvisionTargetVerticalAngleKey = "targetPitch"
+
+kPhotonvisionTargetSimDistanceKey = "simDistance"
```
Comment (lmaxwell24) [inline constants.py:202]: trusting you read the docs right lol

**Review by lmaxwell24 — CHANGES_REQUESTED** [review]

File: commands/reverseintake.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+    # pylint: disable=unused-argument,no-self-use
```
Comment (agelsomini) [inline commands/reverseintake.py:20]: Change this to 
`# pylint: disable-next=unused-argument,no-self-use`

And then add to the following above isFinished
`# pylint: disable-next=no-self-use`

disable-next disables the next line.  Which is better than disable which disables until the end of the block or the end of the file.

File: constants.py [hunk 10]
```diff
@@ -443,3 +466,21 @@
...
+kIntakeCameraMaxHorizontalFOV = Rotation2d.fromDegrees(35)
+kIntakeCameraMinHorizontalFOV = Rotation2d.fromDegrees(-35)
+kIntakeCameraHeightInMeters = 0.5
+kIntakeCameraCenterOffsetInMeters = 0.2
+kIsIntakeCameraCentered = False
+
+kSimIntakeCameraObjectName = "Intake Camera"
+
+# Intake auto intake constants
+kDriveToBallPGain = 0.5
```
Comment (agelsomini) [inline constants.py:482]: Out of curiosity how did these get determined?

Comment (lmaxwell24) [reply constants.py:482]: at the moment its a generic guess - needs to be tuned on the actual system

**Review by lmaxwell24 — APPROVED** [review]

**Comment by agelsomini** [comment]
@lmaxwell24 Can you merge main into this branch.  `git pull origin main` .  I tested and that creates a clean merge commit.  Next PR we can try to do things differently, but that is the fastest way to clean up this branch and make it easy to review the PR.

**Review by lmaxwell24 — DISMISSED** [review]
bias since I also contributed to this commit - will need another peer review from a student

File: constants.py (hunk 8 above)
Comment (LandonBayer) [reply constants.py:212]: At this point we know for sure that we're using two limelight's. This should no longer be photonvision

**Review by LandonBayer — COMMENTED** [review]
Other stuff still applies but these changes are good

File: docs/controllerMappings.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+| ----------------------- | ------ |
+| A                       | 1      |
+| B                       | 2      |
+| X                       | 3      |
+| Y                       | 4      |
+| Left Bumper (LB)        | 5      |
+| Right Bumper (RB)       | 6      |
+| Select                  | 7      |
+| Start                   | 8      |
+| Xbox Button             | 9      |
```
[the remark this answers is not in the document]
Comment (lmaxwell24) [reply docs/controllerMappings.md:17]: xbox button CAN be read, please make sure you are using a wired controller as wireless is different

File: subsystems/indexersubsystem.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+from wpilib import SmartDashboard
+import constants
+from util.helpfulIO import Falcon
+
+
+class IndexerSubsystem(SubsystemBase):
+    def __init__(self) -> None:
+        SubsystemBase.__init__(self)
+        self.setName(__class__.__name__)
+        self.indexerRunning = True
```
[the remark this answers is not in the document]
Comment (lmaxwell24) [reply subsystems/indexersubsystem.py:11]: I would say this would then be another state handled independently when we implement it, like it would update based on if the sensor is there

File: subsystems/intakesubsystem.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+
+class IntakeSubsystem(SubsystemBase):
+    def __init__(self) -> None:
+        SubsystemBase.__init__(self)
+        self.setName(__class__.__name__)
+        self.intakeDeployed = False  # default to intake retracted and not reversed
+        self.intakeReversed = False
+
+        self.pneumaticsHub = PneumaticHub(constants.kPneumaticsHubCanID)
```
Comment (lmaxwell24) [inline subsystems/intakesubsystem.py:14]: unused hub? why does it even need to be initialized if nothing is directly being done with it

**Review by lmaxwell24 — CHANGES_REQUESTED** [review]

File: subsystems/intakesubsystem.py (hunk 13 above)
Comment (LandonBayer) [reply subsystems/intakesubsystem.py:14]: It is being used, the solenoids just default to it so it doesn't have to be called. Pretty sure it's required to be there though (and also for the real robot)

File: subsystems/intakesubsystem.py (hunk 3 above)
Comment (LandonBayer) [reply subsystems/intakesubsystem.py:12]: Did this ever get fixed or no lol

Comment (icai0) [reply subsystems/intakesubsystem.py:12]: yes

**Review by LandonBayer — DISMISSED** [review]

File: subsystems/intakesubsystem.py (hunk 13 above)
Comment (lmaxwell24) [reply subsystems/intakesubsystem.py:14]: if it defaults, it doesnt have to be constructed, I read the docs again (https://robotpy.readthedocs.io/projects/wpilib/en/stable/wpilib/Solenoid.html?highlight=Solenoid) and it doesnt seem to be needed....

**Review by lmaxwell24 — CHANGES_REQUESTED** [review]

File: subsystems/intakesubsystem.py (hunk 13 above)
Comment (LandonBayer) [reply subsystems/intakesubsystem.py:14]: Maybe you're right, it doesn't explicitly say that. I had assumed that it would need to know there is a pneumatics hub to assign solenoids on it. I guess give it a shot without the constructor and see what happens

**Review by lmaxwell24 — DISMISSED** [review]

File: subsystems/indexersubsystem.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+    def reversePath(self) -> None:
+        self.state = self.Mode.Reversed
+
+    def defaultIndexer(self) -> None:
+        if self.state == self.Mode.FeedForward:
+            self.indexerMotor.setSpeed(constants.kIndexerSpeed)
+            self.stagingMotor.setSpeed(constants.kStagingSpeed)
+        elif self.state == self.Mode.Holding:
+            if self.indexerSensor.value() and self.stagingSensor.value():
```
Comment (LandonBayer) [inline subsystems/indexersubsystem.py:72]: Technically this only has to be the indexer sensor because the staging sensor will always be covered when the indexer one is; however, this would not account for the possibility of a passing ball activating it and therefore shutting off the motors, so let's stick with this.

TLDR: No change

File: subsystems/intakesubsystem.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+from wpilib import PneumaticsModuleType, Solenoid, SmartDashboard
+import constants
+from util.helpfulIO import Falcon
+
+
+class IntakeSubsystem(SubsystemBase):
+    def __init__(self) -> None:
+        SubsystemBase.__init__(self)
+        self.setName(__class__.__name__)
+        self.intakeDeployed = False  # default to intake retracted and not reversed
```
Comment (LandonBayer) [inline subsystems/intakesubsystem.py:11]: Feels like intakesubsystem should get the same revamp that indexer did, with enums and stuff to make it more organized. I can do this, but anyone have rejections to that?

File: util/helpfulIO.py [hunk 16]
```diff
@@ -8,19 +7,18 @@
...
 class Falcon:  # represents either a simulated motor or a real Falcon 500
     def __init__(
         self,
         name: str,
         realId: int,
         simId: int,
-        PGain: int = 1,
-        IGain: int = 0,
-        DGain: int = 0,
+        PGain: float = 0.01,
```
Comment (LandonBayer) [inline util/helpfulIO.py:17]: Why so low?

File: subsystems/intakesubsystem.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+    def deployIntake(self) -> None:
+        self.intakeDeployed = True
+        self.intakeReversed = False
+        self.intakeSolenoid.set(True)
+
+    def retractIntake(self) -> None:
+        self.intakeDeployed = False
+        self.intakeSolenoid.set(False)
+        print(
```
Comment (LandonBayer) [inline subsystems/intakesubsystem.py:56]: My crappy debug can also come out after we make sure this works and stuff

**Review by LandonBayer — CHANGES_REQUESTED** [review]

File: util/helpfulIO.py (hunk 16 above)
Comment (lmaxwell24) [reply util/helpfulIO.py:17]: default should be a safe number - P gain of 1 with little strain on the motor causes it to go back and forth

File: subsystems/intakesubsystem.py (hunk 15 above)
Comment (LandonBayer) [reply subsystems/intakesubsystem.py:11]: Alright, I will give it a shot sometime later today based off of indexer's functionality. Shouldn't be to much of a problem as this is my expertise at this point (copying code and changing it slightly to work on other stuff)

File: subsystems/intakesubsystem.py [hunk 18]
```diff
@@ -48,19 +48,9 @@ def unreverseIntake(self) -> None:
     def deployIntake(self) -> None:
         self.intakeDeployed = True
         self.intakeReversed = False
-        self.intakeSolenoid.set(True)
 
     def retractIntake(self) -> None:
         self.intakeDeployed = False
-        self.intakeSolenoid.set(False)
```
Comment (LandonBayer) [inline subsystems/intakesubsystem.py]: Uhh we need that solenoid :)

**Review by LandonBayer — CHANGES_REQUESTED** [review]

File: util/helpfulIO.py (hunk 16 above)
Comment (LandonBayer) [reply util/helpfulIO.py:17]: Alright, my only issue with this is it makes it a bit tougher cuz then we basically need to specify a special P gain for every motor: we should probably do that anyway but for me I don't generally know what the best ones are going to be.

File: robotcontainer.py [hunk 19]
```diff
@@ -110,5 +153,16 @@ def configureButtonBindings(self):
             *self.operatorInterface.driveToTargetControl
         ).whenHeld(DriveToTarget(self.drive, constants.kAutoTargetOffset))
 
+        commands2.button.JoystickButton(
+            *self.operatorInterface.autoBallIntakeControl
+        ).whenHeld(AutoBallIntake(self.drive, self.intake))
+
+        commands2.button.JoystickButton(*self.operatorInterface.shootBall).whenHeld(
+            ShootBall(self.indexer)
+        )
```
Comment (LandonBayer) [inline robotcontainer.py:162]: Why not link this .whenHeld and the next whenRelease on the same button together, like on line 104?

**Review by LandonBayer — APPROVED** [review]

**Review by lmaxwell24 — APPROVED** [review]

**Review by caloisio — APPROVED** [review]
